flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationHandlerImpl.java

```diff
@@ -0,0 +1,297 @@
+                oldState,
+                state,
+                jobId,
+                notFinishedExecutionStates);
+    }
+
+    /**
+     * Handles the termination of the {@code StopWithSavepointTerminationHandler} exceptionally
+     * after triggering a global job fail-over.
+     *
+     * @param completedSavepoint the completed savepoint that needs to be discarded.
+     * @param unfinishedExecutionStates the unfinished states that caused the failure.
+     */
+    private void terminateExceptionallyWithGlobalFailover(
+            CompletedCheckpoint completedSavepoint,
+            Iterable<ExecutionState> unfinishedExecutionStates) {
+        String errorMessage =
+                String.format(
+                        "Inconsistent execution state after stopping with savepoint. At least one execution is still in one of the following states: %s. A global fail-over is triggered to recover the job %s.",
+                        StringUtils.join(unfinishedExecutionStates, ", "), jobId);
+        FlinkException inconsistentFinalStateException = new FlinkException(errorMessage);
+
+        scheduler.handleGlobalFailure(inconsistentFinalStateException);
+        discardSavepoint(completedSavepoint, inconsistentFinalStateException);
+
+        terminateExceptionally(inconsistentFinalStateException);
+    }
+
+    /**
+     * Handles the termination of the {@code StopWithSavepointTerminationHandler} exceptionally
+     * without triggering a global job fail-over. It does restart the checkpoint scheduling.
+     *
+     * @param throwable the error that caused the exceptional termination.
+     */
+    private void terminateExceptionally(Throwable throwable) {
+        checkpointScheduling.startCheckpointScheduler();
+        result.completeExceptionally(throwable);
+    }
+
+    /**
+     * Handles the successful termination of the {@code StopWithSavepointTerminationHandler}.
+     *
+     * @param completedSavepoint the completed savepoint
+     */
+    private void terminateSuccessfully(CompletedCheckpoint completedSavepoint) {
+        result.complete(completedSavepoint.getExternalPointer());
+    }
+
+    private void discardSavepoint(
+            CompletedCheckpoint completedSavepoint,
+            FlinkException inconsistentFinalStateException) {
+        CompletableFuture.runAsync(
+                () -> {
+                    try {
+                        completedSavepoint.discard();
+                    } catch (Exception e) {
+                        log.warn(
+                                "Error occurred while cleaning up completed savepoint due to stop-with-savepoint failure.",
+                                e);
+                        inconsistentFinalStateException.addSuppressed(e);
+                    }
+                },
+                ioExecutor);
```

Comment:
How are we ensuring that this action is executed before we shut down the scheduler?

Reply:
The `ioExecutor` of the `Scheduler` is passed down from the `JobManagerSharedServices.scheduledExecutorService` which has a [shutdown procedure](https://github.com/XComp/flink/blob/2c4e0ab921ccfaf003073ee50faeae4d4e4f4c93/flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerSharedServices.java#L88) for gracefully shutting down the `Executor`. Shouldn't this be independent from the `Scheduler` shutdown? 🤔

Reply:
Well, it depends on whose responsibility it is to clean up the savepoint. The existing shutdown procedure won't guarantee that enqueued tasks are executed. `ExecutorService.shutdownNow` will only execute the currently running tasks and clear all other tasks. Hence, there is no guarantee that the savepoint is actually removed.

Reply:
I went for the easy fix to add a graceful shutdown to the `JobManagerSharedServices` for now after discussing solutions with @zentol today. Here's what we came up with:
The `StopWithSavepointTerminationHandler` interface would need to be adapted: Instead of providing a dedicated method for returning the result, `StopWithSavepointTerminationHandler.handleExecutionsTermination` has to return the `CompletableFuture` result. This enables us to add an optional execution for discarding the savepoint in case of global fail-over. The `StopWithSavepointTerminationManager` would forward this `CompletableFuture` as a result. In case of global fail-over, the result would only complete if the discard call terminated.

The `CompletableFuture` returned by the `StopWithSavepointTerminationManager` could be collected by the `SchedulerBase` now. Multiple calls to `stopWithSavepoint` would result in combining the `CompletableFuture` results into a single `CompletableFuture`. We could then wait in `SchedulerBase.shutDownCheckpointServices` for this `CompletableFuture` to finish. It gets executed when the termination future of the underlying `ExecutionGraph` terminates. The drawback of this approach is that we would block in the main executor which might go against the rule of not having blocking operations on the main executor.
The solution would be to introduce a terminationFuture that only completes if the executions are done and is then handled in `JobMaster.stopScheduling` which is called by `JobMaster.stopJobExecution`.
It felt like it's getting a bit out of scope for the issue here - that's why we decided to go for the graceful shutdown approach for now and discuss the solution first.

Reply:
Yes, to do it properly would mean a lot more work. What we would have to do is to make the `SchedulerNG` asynchronously stoppable. One could for example let it extend `AutoCloseableAsync`. The `JobMaster` would then have to close the scheduler asynchronously by calling the `closeAsync` method and then make its shutdown dependent on the returned future. The future itself should only be completed once all savepoints and potentially other operations have completed. That way we never have to execute anything blocking in the main thread.

Hence, introducing a grace period for shutting down the `ioExecutor` sounds like a good first step.

Reply:
We switched back to not removing the Savepoint in case of fail over due to the reasons mentioned in [this comment](https://github.com/apache/flink/pull/14847#issuecomment-785303451).